rfcs/20210415-gpu-only/README.md

```diff
@@ -0,0 +1,185 @@
+* For example, for `dnnl_set_max_cpu_isa` oneDNN
+can return `runtime_error` because it's a dedicated API for CPU but for
+`dnnl_gemm_s8s8s32` oneDNN can return `unimplemented`.
+* The GTest will be adjusted to make sure that such APIs return proper status
+
+The suggestion is to go with the second option as this is better aligned with the
+runtime agnostic approach that oneDNN uses for almost all APIs. Also, it allows
+applications that use CPU specific API under a runtime condition to link with
+oneDNN.
+
+### Cross-Engine Reorder
+
+Since CPU engine is not available in GPU only configuration the cross-engine
+reorder capability will be limited to GPU -> GPU case. The only exception is when
+a user creates `dnnl::memory` using a pointer (handle) to shared USM. In this case,
+the behavior will be similar to CPU -> GPU and GPU -> CPU cases.
+
+OpenVINO mentioned that they used the cross-engine reorder to avoid dealing with
+runtime directly when there is a need to create `dnnl::memory` for GPU from a
+host pointer. For GPU only configurations users can use `map`/`unmap` functionality
+ to copy host data to GPU.
+
+There is also an option to improve usability for such scenarios by introducing the
+following API.
+
+SYCL interoperability API:
+
+```cpp
+// dnnl_sycl.hpp
+
+namespace dnnl {
+namespace sycl_interop {
+
+/// The two `make_memory` functions can be used to create a memory object that
+/// contains a copy of data from the buffer @p host_ptr is pointing to.
+/// @param host_ptr - is a host accessible pointer
+inline memory make_memory(const memory::desc &memory_desc,
+    const engine &aengine, memory_kind kind,
+    const void *host_ptr, void *handle = DNNL_MEMORY_ALLOCATE);
+
+template <typename T, int ndims = 1>
+memory make_memory(const memory::desc &memory_desc, const engine &aengine,
+    const void *host_ptr, cl::sycl::buffer<T, ndims> &abuffer);
+
+}
+}
+```
+
+OpenCL interoperability API:
+```c
+// dnnl_ocl.h
+
+// Similar two the non-buffer `make_memory` version.
+dnnl_status_t DNNL_API dnnl_ocl_interop_memory_create_from_host_ptr(dnnl_memory_t *memory,
+        const dnnl_memory_desc_t *memory_desc, dnnl_engine_t engine,
+        const void *host_ptr, void *handle);
+```
+
+This API will also be useful if current API will be split in the future because in
+that case cross-reorder cannot be supported as well.
+
+However, this can be postponed until there is a request.
+
+### Threading CPU Runtime
+
+While the CPU runtime is absent there is still a need for CPU threading runtime
+because all threading functionality (e.g `parallel_nd`), which is used in the
+testing, is using it. There are two options to handle threading in the testing:
+
+#### Option 1: Enable CPU Threading Runtime in the Library
+* Enable TBB threading for DPC++ GPU only build
+* Enable OpenMP threading for OpenCL GPU only build
+
+Note: the cases when OpenMP runtime is not available (e.g. for Xcode Clang) should
+fall back to sequential runtime.
```

Comment:
I  wonder if OpenVINO will be fine with that OMP dependency since they use OpenCL and TBB.

Reply:
I asked them for feedback

Reply:
In GPU plugin we use TBB as a default threading backend, so OMP might cause some issues I believe, and introduce redundant dependency. Is it possible to switch to TBB for OCL engine as well?

Reply:
This is possible but it will require users to fulfil TBB dependency even though CPU runtime is absent. While it may be fine for OpenVINO it can be a problem for other users. I think we need to go with the second option then, which is about reusing threading layer in testing but build oneDNN for sequential runtime. In this case oneDNN will not have a dependency on any threading runtime.